justfile

```diff
@@ -0,0 +1,147 @@
+    PYTHON_VERSION=${PYTHON_VERSION:-python3.11}
+
+    # create venv and upgrade pip
+    if [[ ! -d $VIRTUAL_ENV ]]; then
+      $PYTHON_VERSION -m venv $VIRTUAL_ENV
+      $PIP install --upgrade pip
+    fi
+
+
+# run pip-compile with our standard settings
+pip-compile *args: devenv
+    #!/usr/bin/env bash
+    set -euo pipefail
+
+    $BIN/pip-compile --allow-unsafe --generate-hashes --strip-extras {{ args }}
+
+
+# ensure dev and prod requirements installed and up to date
+devenv: virtualenv
+    #!/usr/bin/env bash
+    set -euo pipefail
+
+    for req_file in requirements.dev.txt requirements.prod.txt; do
+      # If we've installed this file before and the original hasn't been
+      # modified since then bail early
+      record_file="$VIRTUAL_ENV/$req_file"
+      if [[ -e "$record_file" && "$record_file" -nt "$req_file" ]]; then
+        continue
+      fi
+
+      if cmp --silent "$req_file" "$record_file"; then
+        # If the timestamp has been changed but not the contents (as can happen
+        # when switching branches) then just update the timestamp
+        touch "$record_file"
+      else
+        # Otherwise actually install the requirements
+
+        # Collapse output when running in Github Actions
+        [[ -v CI ]] && echo "::group::Install $req_file (click to view)" || true
+        # --no-deps is recommended when using hashes, and also works around a
+        # bug with constraints and hashes. See:
+        # https://pip.pypa.io/en/stable/topics/secure-installs/#do-not-use-setuptools-directly
+        $PIP install --no-deps -r "$req_file"
+        [[ -v CI ]]  && echo "::endgroup::" || true
+
+        # Make a record of what we just installed
+        cp "$req_file" "$record_file"
+      fi
+    done
+
+
+# lint and check formatting but don't modify anything
+check: devenv
+    #!/usr/bin/env bash
+
+    failed=0
+
+    check() {
+      # Display the command we're going to run, in bold and with the "$BIN/"
+      # prefix removed if present
+      echo -e "\e[1m=> ${1#"$BIN/"}\e[0m"
+      # Run it
+      eval $1
+      # Increment the counter on failure
+      if [[ $? != 0 ]]; then
+        failed=$((failed + 1))
+        # Add spacing to separate the error output from the next check
+        echo -e "\n"
+      fi
+    }
+
+    check "$BIN/ruff format --diff --quiet ."
+    check "$BIN/ruff check --show-source ."
+    check "docker run --rm -i ghcr.io/hadolint/hadolint:v2.12.0-alpine < Dockerfile"
+
+    if [[ $failed > 0 ]]; then
+      echo -en "\e[1;31m"
+      echo "   $failed checks failed"
+      echo -e "\e[0m"
+      exit 1
+    fi
+
+
+# fix formatting and import sort ordering
+fix: devenv
+    $BIN/ruff format .
+    $BIN/ruff --fix .
+
+
+# run Django's manage.py entrypoint
+manage *ARGS: devenv
+    $BIN/python manage.py {{ ARGS }}
+
+
+# run tests
+test *ARGS: devenv
+    $BIN/python -m pytest {{ ARGS }}
+
+
+# run tests as they will be in run CI (checking code coverage etc)
+@test-all: devenv
```

Comment:
Huh, TIL about `@command` Never had need of it before.

I think I may be less sensitive to excess verbiage that others.

Reply:
This _creates_ more verbiage though, doesn't it? In that it echos the script contents before running it. I didn't come up with the idea (it's borrowed from the ehrQL justfile) but I think it's nice to have the script echoed in the CI output.

Reply:
an `@` in this position stops just from outputting the commandline just before running it

```
test-all:
  @echo "thing"
```

you're saying that an `@` in this position outputs the whole action before running?

```
@test-all:
  echo "thing"
```

Reply:
ah yes https://github.com/casey/just#quiet-recipes

Reply:
Yeah, it's _super_ confusing. A `@` prefix on the recipe makes it "quiet" which inverts the usual echoing behaviour so lines which would normally be echoed are suppressed and lines prefixed with `@` are echoed.

But the default behaviour for shebang recipes is _not_ to echo anything. So by making it "quiet" it ends up echoing the script. Easy, right? :joy: 
https://github.com/casey/just?tab=readme-ov-file#quiet-recipes

Reply:
Huh, I totally read that wrong!